### PR TITLE
added a listener for testing host to test manual initialization on test app

### PR DESCRIPTION
**Pull request by KangxuanYe**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

This is a change on test app that it will not go through fixed initialization flow we hard-coded here if one of url parameter is `testInit`. Instead, it will have a listener listening to our testing host, waiting for message to call corresponding APIs to complete the initialization flow.

### Main changes in the PR:

1. Add a message listener for testing host to call initialization relevant APIs
2. Set a url parameter so that hard-coded initialization flow will not be handled.

### CODE REVIEW

File: apps/teams-test-app/src/App.tsx [hunk 1]
```diff
@@ -16,29 +16,61 @@ const urlParams = new URLSearchParams(window.location.search);
...
-    app.notifyAppLoaded();
-    app.notifySuccess();
+window.addEventListener('message', handleMessageFromMockedHost);
+
+function handleMessageFromMockedHost(msg: MessageEvent): void {
+  if (!msg || !msg.data) {
+    console.log('Unrecognized message format received by app, message being ignored. Message: %o', msg);
+    return;
+  }
+  // Ensure the message is from a trusted origin
```
Comment (TrevorJoelHarris) [inline apps/teams-test-app/src/App.tsx:54]: How would you feel about writing a log statement for the cases where we are handling the message? It might be useful to be able to tell from the console logs that `app.initialize` was called because of this handler as opposed to some other reason.

Maybe the simplest way to do it would be something like 
```suggestion
console.log(`Received message from test host: ${msg.data}`);

  // Ensure the message is from a trusted origin
```

File: apps/teams-test-app/src/App.tsx [hunk 2]
```diff
@@ -16,29 +16,61 @@ const urlParams = new URLSearchParams(window.location.search);
...
-) {
-  console.info('Not calling appInitialization because part of App Initialization Test run');
-} else {
-  if (isTestBackCompat()) {
-    appInitialization.notifyAppLoaded();
-    appInitialization.notifySuccess();
-  } else {
-    app.notifyAppLoaded();
-    app.notifySuccess();
+window.addEventListener('message', handleMessageFromMockedHost);
```
Comment (TrevorJoelHarris) [inline apps/teams-test-app/src/App.tsx:47]: Should this listener only be set up when testInit is present on the url paremeters and set to true?

File: apps/teams-test-app/src/App.tsx [hunk 3]
```diff
@@ -16,29 +16,61 @@ const urlParams = new URLSearchParams(window.location.search);
...
+      break;
+    case 'app.notifyExpectedFailure':
+      app.notifyExpectedFailure({ reason: app.ExpectedFailureReason.Other, message: 'Failed on test app on purpose' });
+      break;
+    case 'app.notifyAppLoaded':
+      app.notifyAppLoaded();
+      break;
+    // Add more cases for other API calls as needed
+    default:
+      console.log('Unknown API call or response:', msg);
```
Comment (TrevorJoelHarris) [inline apps/teams-test-app/src/App.tsx:73]: Should this be console.warn instead of console.log?

File: apps/teams-test-app/src/App.tsx [hunk 4]
```diff
@@ -16,29 +16,61 @@ const urlParams = new URLSearchParams(window.location.search);
...
-    appInitialization.notifyAppLoaded();
-    appInitialization.notifySuccess();
-  } else {
-    app.notifyAppLoaded();
-    app.notifySuccess();
+window.addEventListener('message', handleMessageFromMockedHost);
+
+function handleMessageFromMockedHost(msg: MessageEvent): void {
+  if (!msg || !msg.data) {
+    console.log('Unrecognized message format received by app, message being ignored. Message: %o', msg);
```
Comment (TrevorJoelHarris) [inline apps/teams-test-app/src/App.tsx:51]: Should this be console.warn instead of console.log?

**Review by TrevorJoelHarris — CHANGES_REQUESTED** [review]
Left some thoughts about when to set up the listener

File: apps/teams-test-app/src/App.tsx [hunk 5]
```diff
@@ -16,29 +16,61 @@ const urlParams = new URLSearchParams(window.location.search);
 const getOriginsParam = urlParams.has('origins') && urlParams.get('origins') ? urlParams.get('origins') : '';
 const validMessageOrigins: string[] | undefined = getOriginsParam ? getOriginsParam.split(',') : undefined;
 
-// This is added for custom initialization when app can be initialized based upon a trigger/click.
-if (!urlParams.has('customInit') || !urlParams.get('customInit')) {
-  if (isTestBackCompat()) {
-    initialize(undefined, validMessageOrigins);
+if (!urlParams.has('testInit') || !urlParams.get('testInit')) {
```
Comment (AE-MS) [inline apps/teams-test-app/src/App.tsx:19]: >!urlParams.has('testInit') || !urlParams.get('testInit') [](http://example.com/codeflow?start=4&length=56)

Isn't checking `urlParams.get('testInit') === null` sufficient?

Comment (AE-MS) [inline apps/teams-test-app/src/App.tsx:19]: >testInit [](http://example.com/codeflow?start=20&length=8)

Is it possible to name this parameter in a way that is more description of what it does? I.e. "testInit" doesn't exactly say what behavior is being triggered. Perhaps the parameter could be named something like "doNotCallInitializeFunction" ? I'm not actually sure what this parameter is doing... except telling it to ignore the customInit parameter? Could we just not send the customInit parameter in situations where you don't want the behavior it triggers?

File: apps/teams-test-app/src/App.tsx (hunk 2 above)
Comment (KangxuanYe) [reply apps/teams-test-app/src/App.tsx:47]: Yes, I will put it into the situation when init is skipped.

File: apps/teams-test-app/src/App.tsx (hunk 5 above)
Comment (KangxuanYe) [reply apps/teams-test-app/src/App.tsx:19]: Logically it seems like it doesn't work if there is a url link looks as xxxxxx?testInit=false. And `urlParams.get('testInit') === null` has the same behavior if `testInit=true` or `testInit=false`

Comment (KangxuanYe) [reply apps/teams-test-app/src/App.tsx:19]: There is a possibility I will remove this name because `customInit` and `appInitializationTest` url parameters do the same thing as it does.

File: apps/teams-test-app/src/App.tsx [hunk 6]
```diff
@@ -16,29 +16,61 @@ const urlParams = new URLSearchParams(window.location.search);
...
+  // This is added for custom initialization when app can be initialized based upon a trigger/click.
+  if (!urlParams.has('customInit') || !urlParams.get('customInit')) {
```
[the remark this answers is not in the document]
Comment (KangxuanYe) [reply apps/teams-test-app/src/App.tsx]: Synced offline. Basically, `customInit` and `appInitializationTest` are all used to skip the hardcode initialization flow.

File: apps/teams-test-app/src/App.tsx [hunk 7]
```diff
@@ -42,6 +43,35 @@ if (
   }
 }
 
+function handleMessageFromMockedHost(msg: MessageEvent): void {
+  if (!msg || !msg.data) {
```
Comment (AE-MS) [inline apps/teams-test-app/src/App.tsx:47]: >!msg [](http://example.com/codeflow?start=6&length=4)

Is it necessary to perform this check? When would the event object ever be `undefined` or `null`? The prototype for `addEventListener` doesn't indicate that the event should ever be `undefined`.

File: apps/teams-test-app/src/App.tsx [hunk 8]
```diff
@@ -42,6 +43,35 @@ if (
...
+    console.warn('Unrecognized message format received by app, message being ignored. Message: %o', msg);
+    return;
+  }
+  console.log(`Received message from test host: ${msg.data}`);
+  // Ensure the message is from a trusted origin
```
Comment (AE-MS) [inline apps/teams-test-app/src/App.tsx:52]: Is this comment accurate? I'm not sure I see how the code below this is ensuring the message is coming from a trusted origin.

**Review by AE-MS — CHANGES_REQUESTED** [review]
:clock1:

File: apps/teams-test-app/src/App.tsx (hunk 8 above)
Comment (TrevorJoelHarris) [reply apps/teams-test-app/src/App.tsx:52]: Maybe something like

```suggestion
  // Handle messages that are correctly formatted and for func values we recognize
```

File: apps/teams-test-app/src/App.tsx (hunk 7 above)
Comment (KangxuanYe) [reply apps/teams-test-app/src/App.tsx:47]: My initial thought is that, If I don't add this check, it is to say I need to make sure, or we have an implicit agreement, that our testing host WILL NOT send empty message. It would not happen right now but we cannot make sure we don't want to do it in the future. I don't want this function to rely on what testing host does/sends so I added this check.

But now I think it would be better if it fails directly if people trying to pass in empty msg instead of filter it out by if-else condition check so that developer might not get aware of himself sending empty msg. Will remove it.

File: apps/teams-test-app/src/App.tsx (hunk 8 above)
Comment (KangxuanYe) [reply apps/teams-test-app/src/App.tsx:52]: I think that comment is for user to make sure the message is from a trusted origin rather than saying the functionality of that piece of code lol.

Thank you @TrevorJoelHarris , I have taken your suggestion.

**Review by AE-MS — APPROVED** [review]
:shipit:

**Review by TrevorJoelHarris — APPROVED** [review]
